- [C011] fix method save #2550

uzi18 opened the pull request:
@TD-er tested and ready to merge, it fix method save #2550

Fixes #2550

uzi18 committed:
- [C011] fix method save #2550

TD-er commented:
Great :)

Can you also add some form function to allow setting the number of lines of a form?
The changes I made also reduced the field for the body from N lines to 1 which makes editing it a bit difficult.
And I guess we can use multi-line forms in more places.

uzi18 commented:
@TD-er more hints? I'm not web dev ;)

TD-er commented:
Ah, well I can add one later this evening.
It was part of the original code, which I replaced.

uzi18 committed:
[DHT] change pin into input before delay

uzi18 commented:
@TD-er will leave it for You, have no time now, added also commit for DHT maybe order mathers for dht11, for dht22 works same

uzi18 committed:
[DHT] change to input with pullup

TD-er committed:
Add TextArea option for C11 and fix some build inconsistencies

TD-er commented:
I added some changes to this PR.
- Some build inconsistencies in build definitions
- Moved several convenience functions to just a single forward declaration with default variable values.
- Added TextArea to allow larger text boxes, like the header and body of C011 controller.

I also tested with compiler flags to reduce the size.
These are still present, but I may remove them tomorrow.
Just let Travis build it and see what the build sizes are on the Travis environment.

TD-er committed:
[MemAnalyzer] Fix MemAnalyzer.py script to work with USES_.... defines

The MemAnalyzer tool was not able to perform a proper analysis of the current builds, since we introduced the `USES_P001` syntax of defines to determine what should be included and what not.

TD-er commented:
I've been looking into a lot of options to reduce the size of the binaries.
For analysis purposes, I fixed the memanalyzer.py tool we had once running.

It is still running here, but for reference the output of the controllers:

module                        |cache IRAM |init RAM   |r.o. RAM   |uninit RAM |Flash ROM
---                           | ---       | ---       | ---       | ---       | ---
CORE                          |29268      |1588       |3012       |35840      |592612
src/_C001.ino                 |0          |0          |4          |16         |4576
src/_C002.ino                 |0          |76         |36         |1136       |18272
src/_C003.ino                 |0          |0          |4          |16         |2336
src/_C003.ino                 |0          |0          |4          |16         |2336
src/_C004.ino                 |0          |0          |0          |16         |2464
src/_C005.ino                 |0          |4          |0          |1136       |10048
src/_C006.ino                 |0          |4          |0          |1136       |10096
src/_C007.ino                 |0          |0          |0          |16         |2592
src/_C008.ino                 |0          |0          |0          |16         |4320
src/_C009.ino                 |0          |72         |36         |16         |6576
src/_C010.ino                 |0          |0          |0          |16         |3552
src/_C011.ino                 |0          |0          |0          |16         |6144
src/_C012.ino                 |0          |0          |0          |16         |4000
src/_C013.ino                 |0          |0          |0          |16         |1296
src/_C014.ino                 |0          |4          |324        |1136       |19392
src/_C015.ino                 |0          |0          |1144       |120        |7664
src/_C016.ino                 |0          |0          |0          |32         |5056
src/_C017.ino                 |0          |72         |40         |24         |4592
src/_C018.ino                 |756        |0          |112        |80         |19952

uzi18 commented:
maybe we can drop some bits from core - 600k?

TD-er commented:
You also never sleep, do you ;)

Well one thing I found out is that we need to move stuff around for various reasons.

There is a bug in the way the .ino.cpp file is created.
This means when the forward declaration is generated of a function in a .ino file, the surrounding #ifdef markings is not included.
This also explains our issue from last week to have the structs for the controllers queue removed.
Also the issues seen by @clumsy-stefan can be explained with this.

Another thing is the way we do things in .h files.
Strictly speaking, we should only have stuff in .h files for declaration and no compiled code.
This means:
- Do as little includes as possible in .h files (use forward declaration where possible)
- Add .cpp files for the functions defined in .h files and have included in .cpp files.

This allows for cleaner code, less include dependencies and easier to wrap parts in #ifdef USES_... wrappers.
Also if all compiled code is in .cpp files, it will only be added if the accompanied .h file is actually being used.
Another thing is that compiled code in .h files is compiled as inline code, so it may lead to bigger binaries. I don't think it is really this much, since we mostly have member functions of a struct in .h files, but you never know.
Especially the code used in controller struct is being used in more places.

One thing we have to look into is, how can this being moved in a directory structure so the ArduinoIDE is also capable of compiling it and don't end up with lots and lots of files in 1 directory.

The plugins/controllers can be in .ino files.

TD-er committed:
Disable Home MQTT for dev build to make it fit max sketch size

Is a temporary disable to let the build succeed again.
This controller is included in the test build.

clumsy-stefan commented:
@TD-er it seems or could be that ArduinoIDE handels the described things differently than PIO. That's probably why I could compile and use some changes while PIO couldn't...

uzi18 commented:
@TD-er yes, it was a hard night for me, but usually, prefer to code/whatever at deep night when all of childrens sleeping.

It is something with build process, in my projects usually I use one ino file and all rest is in header files, here this bad behavior shoud be fixed in platformio-esp8266/arduino, maybe it is already adressed in issues?

when it will be addressed/resolved - we can think about link time optimization - lto

uzi18 commented:
@ivankravets maybe you are able to hint us here? https://github.com/letscontrolit/ESPEasy/pull/2582#issuecomment-526961009

TD-er commented:
> @ivankravets maybe you are able to hint us here? [#2582 (comment)](https://github.com/letscontrolit/ESPEasy/pull/2582#issuecomment-526961009)

I already did start an issue about that in the Arduino/esp8266 repo: https://github.com/esp8266/Arduino/issues/6475#issuecomment-526914526

uzi18 commented:
@TD-er thx for reference

have already start digging where ino.cpp is created but still not found

TD-er commented:
I think it is best to merge this PR (and the other for changing stuff in platformio.ini) and then start moving stuff around based on the experience I described earlier.

uzi18 commented:
maybe if you leave only espeasy.ino from espeasy core, and move others to .h files with one include in top of espeasy.ino ... we can hit goal, try

TD-er committed:
[Build] Start making Vagrant provisioning for quick self builds

Installing Vagrant:

Vagrantup.com site: https://www.vagrantup.com/docs/installation/
VirtualBox: https://www.virtualbox.org/

Great tutorial: https://www.sitepoint.com/getting-started-vagrant-windows/



After installing Vagrant and VirtualBox, open a shell in the tools/vagrant folder.
Run:
  vagrant up

To halt the VM, type:
  vagrant halt
